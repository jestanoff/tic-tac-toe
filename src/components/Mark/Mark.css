.container {
  display: inline-block;
  overflow: hidden;
  padding-bottom: 100%;
  position: relative;
  text-align: center;
  width: 100%;
}

.svg {
  display: inline-block;
  left: 0;
  position: absolute;
  top: 0;
}
