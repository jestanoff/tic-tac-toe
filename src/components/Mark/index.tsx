import * as React from 'react';
import { DARK_GRAY, SYMBOLS, WHITE } from '../../constants';
import styles from './Mark.css';

type MarkProps = {
  isAnimated?: boolean;
  overrideColor?: string;
  type: string;
};

const Mark = ({ isAnimated = false, overrideColor, type }: MarkProps) => (
  <div className={styles.container}>
    {type === SYMBOLS[1] && (
      <svg
        className={styles.svg}
        data-testid='cross-mark'
        preserveAspectRatio='xMinYMin meet'
        version='1.1'
        viewBox='0 0 500 500'
      >
        <line x1='130' y1='130' x2='130' y2='130' strokeWidth='50' stroke={overrideColor ?? DARK_GRAY}>
          <animate
            attributeName='x2'
            attributeType='XML'
            dur={isAnimated ? '125ms' : '1ms'}
            fill='freeze'
            from='130'
            id='first'
            repeatCount='1'
            to='370'
          />
          <animate
            attributeName='y2'
            attributeType='XML'
            dur={isAnimated ? '125ms' : '1ms'}
            fill='freeze'
            from='130'
            repeatCount='1'
            to='370'
          />
        </line>
        <line
          stroke={overrideColor || DARK_GRAY} opacity='0'
          strokeWidth='50'
          x1='370'
          x2='370'
          y1='130'
          y2='130'
        >
          <animate
            attributeName='opacity'
            attributeType='XML'
            begin={isAnimated ? '125ms' : '0s'}
            dur={isAnimated ? '125ms' : '1ms'}
            fill='freeze'
            from='0'
            repeatCount='1'
            to='1'
          />
          <animate
            attributeName='x2'
            attributeType='XML'
            begin={isAnimated ? '125ms' : '0s'}
            dur={isAnimated ? '125ms' : '1ms'}
            fill='freeze'
            from='370'
            repeatCount='1'
            to='130'
          />
          <animate
            attributeName='y2'
            attributeType='XML'
            begin={isAnimated ? '125ms' : '0s'}
            dur={isAnimated ? '125ms' : '1ms'}
            fill='freeze'
            from='130'
            repeatCount='1'
            to='370'
          />
        </line>
      </svg>
    )}
    {type === SYMBOLS[2] && (
      <svg
        className={styles.svg}
        data-testid='circle-mark'
        preserveAspectRatio='xMinYMin meet'
        version='1.1'
        viewBox='0 0 500 500'
      >
        <circle
          cx='250'
          cy='250'
          r='125'
          stroke={overrideColor || WHITE}
          strokeWidth='45'
          fill='transparent'
          strokeDasharray='800'
          strokeDashoffset='-800'
        >
          <animate
            attributeType='XML'
            attributeName='stroke-dashoffset'
            from='-800'
            to='0'
            dur={isAnimated ? '250ms' : '1ms'}
            repeatCount='1'
            fill='freeze'
          />
        </circle>
      </svg>
    )}
  </div>
);

export default React.memo(Mark);
