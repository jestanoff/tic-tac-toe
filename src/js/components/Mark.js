import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styles from '../../css/mark.css';
import { DARK_GRAY, WHITE } from '../constants';

const Mark = ({ isAnimated, overrideColor, type }) => (
  <div className={styles.container}>
    {type === 'cross' ? (
      <svg className={styles.svg} preserveAspectRatio="xMinYMin meet" version="1.1" viewBox="0 0 500 500">
        <line x1="130" y1="130" x2="130" y2="130" strokeWidth="50" stroke={overrideColor || DARK_GRAY}>
          <animate
            attributeName="x2"
            dur={isAnimated ? '125ms' : '1ms'}
            fill="freeze"
            from="130"
            id="first"
            repeatCount="1"
            to="370"
          />
          <animate
            attributeName="y2"
            dur={isAnimated ? '125ms' : '1ms'}
            fill="freeze"
            from="130"
            repeatCount="1"
            to="370"
          />
        </line>
        <line x1="370" y1="130" x2="370" y2="130" strokeWidth="50" stroke={overrideColor || DARK_GRAY} opacity="0">
          <animate
            attributeName="opacity"
            begin={isAnimated ? '125ms' : '0s'}
            dur="1ms"
            fill="freeze"
            from="0"
            repeatCount="1"
            to="1"
          />
          <animate
            attributeName="x2"
            begin={isAnimated ? '125ms' : '0s'}
            dur={isAnimated ? '125ms' : '1ms'}
            fill="freeze"
            from="370"
            repeatCount="1"
            to="130"
          />
          <animate
            attributeName="y2"
            begin={isAnimated ? '125ms' : '0s'}
            dur={isAnimated ? '125ms' : '1ms'}
            fill="freeze"
            from="130"
            repeatCount="1"
            to="370"
          />
        </line>
      </svg>
    ) : (
      <svg version="1.1" viewBox="0 0 500 500" preserveAspectRatio="xMinYMin meet" className={styles.svg}>
        <circle
          cx="250"
          cy="250"
          fill="transparent"
          r="125"
          stroke={overrideColor || WHITE}
          strokeDasharray="800"
          strokeDashoffset="-800"
          strokeWidth="45"
        >
          <animate
            attributeName="stroke-dashoffset"
            dur={isAnimated ? '250ms' : '1ms'}
            fill="freeze"
            from="-800"
            repeatCount="1"
            to="0"
          />
        </circle>
      </svg>
    )}
  </div>
);

Mark.propTypes = {
  isAnimated: PropTypes.bool,
  overrideColor: PropTypes.string,
  type: PropTypes.string.isRequired,
};

Mark.defaultProps = {
  isAnimated: false,
  overrideColor: undefined,
};

export default memo(Mark);
