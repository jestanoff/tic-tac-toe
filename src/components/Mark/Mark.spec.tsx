import * as React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Mark from './index';

describe('<Mark />', () => {
  test('should render cross mark as SVG', () => {
    render(<Mark type='cross' />);

    expect(screen.getByTestId('cross-mark')).toBeInTheDocument();
    expect(screen.queryByTestId('circle-mark')).not.toBeInTheDocument();
    expect(screen.getByTestId('cross-mark')).toMatchSnapshot();
  });

  test('should render circle mark as SVG', () => {
    render(<Mark type='circle' />);

    expect(screen.getByTestId('circle-mark')).toBeInTheDocument();
    expect(screen.queryByTestId('cross-mark')).not.toBeInTheDocument();
    expect(screen.getByTestId('circle-mark')).toMatchSnapshot();
  });

  test('should add animation to circle mark', () => {
    render(<Mark isAnimated type='circle' />);

    expect(screen.getByTestId('circle-mark').children[0].children[0].getAttribute('dur')).toBe('250ms');
  });

  test('should not add animation to circle mark', () => {
    render(<Mark isAnimated={false} type='circle' />);

    expect(screen.getByTestId('circle-mark').children[0].children[0].getAttribute('dur')).toBe('1ms');
  });

  test('should add animation to cross mark', () => {
    render(<Mark isAnimated type='cross' />);
    const mark = screen.getByTestId('cross-mark');

    expect(mark.children[0].children[0].getAttribute('dur')).toBe('125ms');
    expect(mark.children[0].children[1].getAttribute('dur')).toBe('125ms');
    expect(mark.children[1].children[0].getAttribute('dur')).toBe('125ms');
    expect(mark.children[1].children[0].getAttribute('begin')).toBe('125ms');
    expect(mark.children[1].children[1].getAttribute('dur')).toBe('125ms');
    expect(mark.children[1].children[1].getAttribute('begin')).toBe('125ms');
    expect(mark.children[1].children[2].getAttribute('dur')).toBe('125ms');
    expect(mark.children[1].children[2].getAttribute('begin')).toBe('125ms');
  });

  test('should not add animation to cross mark', () => {
    render(<Mark isAnimated={false} type='cross' />);
    const mark = screen.getByTestId('cross-mark');

    expect(mark.children[0].children[0].getAttribute('dur')).toBe('1ms');
    expect(mark.children[0].children[1].getAttribute('dur')).toBe('1ms');
    expect(mark.children[1].children[0].getAttribute('dur')).toBe('1ms');
    expect(mark.children[1].children[0].getAttribute('begin')).toBe('0s');
    expect(mark.children[1].children[1].getAttribute('dur')).toBe('1ms');
    expect(mark.children[1].children[1].getAttribute('begin')).toBe('0s');
    expect(mark.children[1].children[2].getAttribute('dur')).toBe('1ms');
    expect(mark.children[1].children[2].getAttribute('begin')).toBe('0s');
  });
});
