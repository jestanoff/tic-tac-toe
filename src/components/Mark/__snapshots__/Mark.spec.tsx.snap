// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Mark /> should render circle mark as SVG 1`] = `
<svg
  class="svg"
  data-testid="circle-mark"
  preserveAspectRatio="xMinYMin meet"
  version="1.1"
  viewBox="0 0 500 500"
>
  <circle
    cx="250"
    cy="250"
    fill="transparent"
    r="125"
    stroke="#FFFFFF"
    stroke-dasharray="800"
    stroke-dashoffset="-800"
    stroke-width="45"
  >
    <animate
      attributeName="stroke-dashoffset"
      attributeType="XML"
      dur="1ms"
      fill="freeze"
      from="-800"
      repeatCount="1"
      to="0"
    />
  </circle>
</svg>
`;

exports[`<Mark /> should render cross mark as SVG 1`] = `
<svg
  class="svg"
  data-testid="cross-mark"
  preserveAspectRatio="xMinYMin meet"
  version="1.1"
  viewBox="0 0 500 500"
>
  <line
    stroke="#444444"
    stroke-width="50"
    x1="130"
    x2="130"
    y1="130"
    y2="130"
  >
    <animate
      attributeName="x2"
      attributeType="XML"
      dur="1ms"
      fill="freeze"
      from="130"
      id="first"
      repeatCount="1"
      to="370"
    />
    <animate
      attributeName="y2"
      attributeType="XML"
      dur="1ms"
      fill="freeze"
      from="130"
      repeatCount="1"
      to="370"
    />
  </line>
  <line
    opacity="0"
    stroke="#444444"
    stroke-width="50"
    x1="370"
    x2="370"
    y1="130"
    y2="130"
  >
    <animate
      attributeName="opacity"
      attributeType="XML"
      begin="0s"
      dur="1ms"
      fill="freeze"
      from="0"
      repeatCount="1"
      to="1"
    />
    <animate
      attributeName="x2"
      attributeType="XML"
      begin="0s"
      dur="1ms"
      fill="freeze"
      from="370"
      repeatCount="1"
      to="130"
    />
    <animate
      attributeName="y2"
      attributeType="XML"
      begin="0s"
      dur="1ms"
      fill="freeze"
      from="130"
      repeatCount="1"
      to="370"
    />
  </line>
</svg>
`;
